ratis-logservice/src/main/java/org/apache/ratis/logservice/api/LogStream.java

```diff
@@ -0,0 +1,109 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.ratis.logservice.api;
+
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * A distributed log with "infinite" length that supports reads and writes.
+ */
+public interface LogStream {
+
+  /**
+   * Returns the unique name to identify this log.
+   */
+  LogName getName();
+
+  /**
+   * Returns the size of this LogStream in bytes.
+   */
+  long getSizeInBytes();
+
+  /**
+   * Returns the number of records in this LogStream.
+   */
+  long getSizeInRecords();
+
+  /**
+   * Creates a reader to read this LogStream asynchronously.
+   *
+   * @return An asynchronous reader
+   */
+  AsyncLogReader createAsyncReader();
+
+  /**
+   * Creates a writer to write to this LogStream asynchronously.
+   *
+   * @return An asynchronous writer
+   */
+  AsyncLogWriter createAsyncWriter();
+
+  /**
+   * Creates a reader to read this LogStream synchronously.
+   *
+   * @return A synchronous reader
+   */
+  LogReader createReader();
+
+  /**
+   * Creates a write to write to this LogStream synchronously.
+   *
+   * @return A synchronous writer
+   */
+  LogWriter createWriter();
+
```

Comment:
I believe the plan is to support a single writer(otherwise commit() would get tedious to manage at server), right ? can we make it getWriter()?

Reply:
My hope was that we could support multiple, independent writers. The writer implementation would be able to hide most details (I hope).

Reply:
Multiple writers - multiple readers is a MUST for generic Log Service (not HBase), but close() must be provided. On commit() - I would replace it with sync() (sync to durable storage). Commit implies rollback :)